### PR TITLE
Generate work status conditions from manifest conditions

### PR DESCRIPTION
**Pull request by elgnay**
This is a followup of pr #4. It aggregates manifest conditions and generates work status conditions.

### CODE REVIEW

File: pkg/spoke/controllers/manifestwork_controller.go [hunk 1]
```diff
@@ -295,7 +295,39 @@ func (m *ManifestWorkController) applyUnstructrued(data []byte, recorder events.
...
+			return nil
+		}
+
+		// aggregate manifest condition to generate work condition
+		applied := 0
+		for _, manifest := range manifestStatus.Manifests {
+			for _, condition := range manifest.Conditions {
+				switch {
+				case condition.Type == string(workapiv1.ManifestApplied) && condition.Status == metav1.ConditionTrue:
+					applied++
```
Comment (qiujian16) [inline pkg/spoke/controllers/manifestwork_controller.go]: i think it needs comment here indicating how to decide the condition of manifestwork. Maybe a separate func would be better. We can reuse that func for all different condition types.  And having a counter seems not necessary. You just need a bool, and set it false if any condition status is false

File: pkg/spoke/controllers/manifestwork_controller_test.go [hunk 2]
```diff
@@ -299,6 +301,26 @@ func (t *testCase) validate(
...
+
+func newManifestCondition(ordinal int32, resource string, conds ...workapiv1.StatusCondition) workapiv1.ManifestCondition {
+	return workapiv1.ManifestCondition{
+		ResourceMeta: workapiv1.ManifestResourceMeta{Ordinal: ordinal, Resource: resource},
+		Conditions:   conds,
+	}
+}
+
 // TestSync test cases when running sync
 func TestSync(t *testing.T) {
```
Comment (qiujian16) [inline pkg/spoke/controllers/manifestwork_controller_test.go]: i think  you need to add the check of the  manifestwork conditions after sync in this test

File: pkg/helper/helper_test.go [hunk 3]
```diff
@@ -187,3 +187,66 @@ func TestSetManifestCondition(t *testing.T) {
...
+				newCondition("one", "False", "my-reason", "my-message", nil),
+				newCondition("two", "True", "my-reason", "my-message", nil),
+			},
+			expectedConditions: []workapiv1.StatusCondition{
+				newCondition("one", "False", "my-reason", "my-message", nil),
+				newCondition("two", "True", "my-reason", "my-message", nil),
+			},
+		},
+		{
+			name: "remove useless status condition",
```
Comment (pmorie) [inline pkg/helper/helper_test.go]: nit: I think `remove old` would be easier to understand than `remove useless`

File: pkg/helper/helper_test.go [hunk 4]
```diff
@@ -187,3 +187,66 @@ func TestSetManifestCondition(t *testing.T) {
...
+				newCondition("two", "True", "my-reason", "my-message", nil),
+			},
+			newConditions: []workapiv1.StatusCondition{
+				newCondition("one", "False", "my-reason", "my-message", nil),
+			},
+			expectedConditions: []workapiv1.StatusCondition{
+				newCondition("one", "False", "my-reason", "my-message", &transitionTime),
+			},
+		},
+	}
```
Comment (pmorie) [inline pkg/helper/helper_test.go:236]: Add a test case where the condition status doesn't change and transition time isn't updated (you could just add this to one of the existing cases too)

File: pkg/spoke/controllers/manifestwork_controller.go [hunk 5]
```diff
@@ -170,7 +170,7 @@ func (m *ManifestWorkController) sync(ctx context.Context, controllerContext fac
 	}
 	if len(errs) > 0 {
 		err = utilerrors.NewAggregate(errs)
-		klog.Errorf("Reconcile work %s fails with err %v: ", manifestWorkName, err)
+		klog.Errorf("Reconcile work %s fails with err: %v", manifestWorkName, err)
```
Comment (pmorie) [inline pkg/spoke/controllers/manifestwork_controller.go:173]: :+1:

File: pkg/spoke/controllers/manifestwork_controller.go [hunk 6]
```diff
@@ -348,3 +373,21 @@ func isSameUnstructured(obj1, obj2 *unstructured.Unstructured) bool {
 
 	return equality.Semantic.DeepEqual(obj1Copy.Object, obj2Copy.Object)
 }
+
+// isAllInCondition checks status of conditions with a particular type in ManifestCondition array.
+// Return true, true only if conditions with the condition type exist and they are all in condition.
+func isAllInCondition(conditionType string, manifests []workapiv1.ManifestCondition) (inCondition bool, exists bool) {
```
Comment (pmorie) [inline pkg/spoke/controllers/manifestwork_controller.go]: nit: `allInCondition` as a name reads a little easier to me than `isAllInCondition`

File: pkg/helper/helper_test.go (hunk 4 above)
Comment (elgnay) [reply pkg/helper/helper_test.go:236]: I think this is already covered by test case 'remove useless status condition'.  Status of condition `one` doesn't change and its transition time is not updated after merge is done.

**Comment by pmorie** [comment]
rebase needed, lgtm otherwise

**Pull request merged**